packages/ses/src/module-load.js

```diff
@@ -0,0 +1,102 @@
+const { create, keys, values, freeze } = Object;
+
+// makeAlias constructs compartment specifier tuples for the `aliases` private
+// field of compartments.
+// These aliases allow a compartment to alias an internal module specifier to a
+// module specifier in an external compartment, and also to create internal
+// aliases.
+// Both are facilitated by the moduleMap Compartment constructor option.
+export const makeAlias = (compartment, specifier) =>
+  freeze({
+    compartment,
+    specifier,
+  });
+
+// Pre-compute resolutions of all imports within the compartment in which a
+// module was loaded.
+const resolveAll = (imports, resolveHook, fullReferrerSpecifier) => {
+  const resolvedImports = create(null);
+  for (const importSpecifier of keys(imports)) {
+    const fullSpecifier = resolveHook(importSpecifier, fullReferrerSpecifier);
+    resolvedImports[importSpecifier] = fullSpecifier;
+  }
+  return freeze(resolvedImports);
+};
+
+// load asynchronously loads ModuleStaticRecords and creates a complete graph
+// of ModuleCompartmentRecords.
+// The module records refer to each other by a reference to the dependency's
+// compartment and the specifier of the module within its own compartment.
+// This graph is then ready to be synchronously linked and executed.
+export const load = async (
+  compartmentPrivateFields,
+  moduleAnalyses,
+  compartment,
+  moduleSpecifier,
+) => {
+  const {
+    resolveHook,
+    importHook,
+    aliases,
+    moduleRecords,
+  } = compartmentPrivateFields.get(compartment);
+
+  // Follow aliases.
+  if (aliases.has(moduleSpecifier)) {
+    const alias = aliases.get(moduleSpecifier);
+    const moduleRecord = await load(
+      compartmentPrivateFields,
+      moduleAnalyses,
+      alias.compartment,
+      alias.specifier,
+    );
+    moduleRecords.set(moduleSpecifier, moduleRecord);
+    return moduleRecord;
+  }
+
+  // Memoize.
+  if (moduleRecords.has(moduleSpecifier)) {
+    return moduleRecords.get(moduleSpecifier);
+  }
+
+  const moduleStaticRecord = await importHook(moduleSpecifier);
+
+  // Guard against invalid importHook behavior.
+  if (!moduleAnalyses.has(moduleStaticRecord)) {
+    throw new TypeError(
+      `importHook must return a ModuleStaticRecord constructed within the same Compartment and Realm`,
+    );
+  }
+
+  const analysis = moduleAnalyses.get(moduleStaticRecord);
+
+  // resolve all imports relative to this referrer module.
+  const resolvedImports = resolveAll(
+    analysis.imports,
+    resolveHook,
+    moduleSpecifier,
+  );
+  const moduleRecord = freeze({
+    ...analysis,
+    compartment,
+    moduleSpecifier,
+    resolvedImports,
+  });
+
+  // Memoize.
+  moduleRecords.set(moduleSpecifier, moduleRecord);
+
+  // Await but do not return the promise for the array of undefineds.
```

Comment:
Why would this (unobserved) value be a promise for an array of `undefined`s? The argument to `Promise.all` is a `map` gather the results of calls to `load`, each of which returns a promise for a `moduleRecord`. To `Promise.all` should return a promise for an array of `moduleRecord`s. Am I missing something?

Reply:
Ah, you are correct. Farther down the stack of upcoming changes, `load` is ultimately called in one place and ignores the resolution. I’ll remove the last line of `load` since it’s superfluous. `load` only needs to block progress toward `importNow` until the graph is fully loaded.

Reply:
gotcha

Reply:
Ah, I was mistaken. `load` is also called here to follow aliases and does need the returned module compartment record for that case. I’ve fixed the comment.